pennylane/gradients/finite_difference.py

```diff
@@ -0,0 +1,223 @@
+        order (int): The order of the finite difference method to use.
+        n (int): compute the :math:`n`-th derivative
+        form (str): The form of the finite difference method. Must be one of
+            ``"forward"``, ``"center"``, or ``"backward"``.
+
+    Returns:
+        tuple[list[QuantumTape], function]: A tuple containing a
+        list of generated tapes, in addition to a post-processing
+        function to be applied to the evaluated tapes.
+
+    **Example**
+
+    >>> with qml.tape.QuantumTape() as tape:
+    ...     qml.RX(params[0], wires=0)
+    ...     qml.RY(params[1], wires=0)
+    ...     qml.RX(params[2], wires=0)
+    ...     qml.expval(qml.PauliZ(0))
+    ...     qml.var(qml.PauliZ(0))
+    >>> tape.trainable_params = {0, 1, 2}
+    >>> gradient_tapes, fn = qml.gradients.finite_diff.grad(tape)
+    >>> res = dev.batch_execute(gradient_tapes)
+    >>> fn(res)
+    [[-0.38751721 -0.18884787 -0.38355704]
+     [ 0.69916862  0.34072424  0.69202359]]
+    """
+    # TODO: replace the JacobianTape._grad_method_validation
+    # functionality before deprecation.
+    diff_methods = tape._grad_method_validation("numeric")
+
+    if not tape.trainable_params or all(g == "0" for g in diff_methods):
+        # Either all parameters have grad method 0, or there are no trainable
+        # parameters.
+        return [], lambda x: np.zeros([tape.output_dim, len(tape.trainable_params)])
+
+    gradient_tapes = []
+    shapes = []
+    c0 = None
+
+    coeffs, shifts = finite_diff_stencil(n, order, form)
+
+    if 0 in shifts:
+        c0 = coeffs[0]
+        gradient_tapes.append(tape)
+        shifts = shifts[1:]
+        coeffs = coeffs[1:]
+
+    # TODO: replace the JacobianTape._choose_params_with_methods
+    # functionality before deprecation.
+    for t_idx, dm in tape._choose_params_with_methods(diff_methods, argnum):
+        if dm == "0":
+            shapes.append(0)
+            continue
+
+        g_tapes = generate_shifted_tapes(tape, t_idx, shifts * h)
+        gradient_tapes.extend(g_tapes)
+        shapes.append(len(g_tapes))
+
+    def processing_fn(results):
+        grads = []
+        start = 1 if c0 is not None else 0
+
+        for s in shapes:
+
+            if s == 0:
+                g = qml.math.convert_like(np.zeros([tape.output_dim]), results)
+                grads.append(g)
+                continue
+
+            res = results[start : start + s]
+            start = start + s
+
+            res = qml.math.stack(res)
+            g = sum([c * r for c, r in zip(coeffs, res)])
+
+            if c0 is not None:
+                g = g + c0 * results[0]
+
+            grads.append(g / (h ** n))
+
+        # The following is for backwards compatibility; currently,
+        # the device stacks multiple measurement arrays, even if not the same
+        # size, resulting in a ragged array.
+        # In the future, we might want to change this so that only tuples
+        # of arrays are returned.

```

Comment:
:100:

Reply:
Shall we do this sooner rather than later?

Reply:
Maybe we can earmark a release for all the breaking changes, so users do not get them bit by bit?